lib/app_profiler/sampler/config.rb

```diff
@@ -0,0 +1,134 @@
+# frozen_string_literal: true
+
+module AppProfiler
+  module Sampler
+    class << self
+      def profile_params(request, config)
+        random = Kernel.rand
+        return unless sample?(random, config, request)
+
+        backend = options_probability(config.backend_options, random)
+        mode = select_mode(backend, options_probability(config.mode_options, random))
+
+        AppProfiler::Parameters.new(
+          mode: mode,
+          interval: config.send("#{mode}_interval"),
+          async: true,
+          backend: backend,
+        )
+      end
+
+      private
+
+      def sample?(random, config, request)
+        return false if random > config.sample_rate
+
+        path = request.path
+        return false unless config.paths.any? { |p| path.match?(p) }
+
+        true
+      end
+
+      # Given options with probabilities, select one based on range.
+      # For example, given options {a: 0.1, b: 0.2, c: 0.7} and random 0.5,
+      # it will return :c
+      # Assumes all probabilities sum to 1
+
+      def options_probability(options, random)
+        current = 0
+        options = options.sort_by do |_, probability|
+          probability
+        end
+
+        options.each do |o, probabilty|
+          current += probabilty
+          if random <= current
+            return o
+          end
+        end
+      end
+
+      def select_mode(backend, mode)
+        return mode if mode == :wall
+        return :retained if backend == :vernier
+
+        :object
+      end
+    end
+
+    class Config
+      attr_reader :sample_rate,
+        :paths,
+        :cpu_interval,
+        :wall_interval,
+        :object_interval,
+        :retained_interval,
+        :wall_mode_probability,
+        :object_or_retained_mode_probability,
+        :stackprof_probability,
+        :vernier_probability
+
+      ## Default values
+      SAMPLE_RATE = 0.001 # 0.1%
+      PATHS = ["/"]
+
+      WALL_INTERVAL = 5000
+      RETAINED_INTERVAL = 0
+      OBJECT_INTERVAL = 1000
+
+      WALL_MODE_PROBABILITY = 0.8 # 80.0%
+      OBJECT_OR_RETAINED_MODE_PROBABILITY = 0.2 # 20.0%
```

Comment:
Not sure I like this, we might see something more analogous to object mode in the future with vernier.

Also, it is possible for wall mode profiles to sample allocations similar to object mode in recent versions.